service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/CbsReceiverLinkHandler.java

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.azure.sdk.iot.service.transport.amqps;
+
+import com.microsoft.azure.sdk.iot.deps.transport.amqp.AmqpsMessage;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.qpid.proton.amqp.messaging.Properties;
+import org.apache.qpid.proton.amqp.messaging.Released;
+import org.apache.qpid.proton.amqp.transport.DeliveryState;
+import org.apache.qpid.proton.engine.Delivery;
+import org.apache.qpid.proton.engine.Event;
+import org.apache.qpid.proton.engine.Receiver;
+
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Every SAS token based authentication over AMQP requires a CBS session with a sender and receiver link. This
+ * class defines the receiver link which receives authentication status codes corresponding to each authentication attempt.
+ */
+@Slf4j
+public final class CbsReceiverLinkHandler extends ReceiverLinkHandler
+{
+    private final static String APPLICATION_PROPERTY_STATUS_CODE = "status-code";
+    private final static String APPLICATION_PROPERTY_STATUS_DESCRIPTION = "status-description";
+
+    private static final String RECEIVER_LINK_ENDPOINT_PATH = "$cbs";
+    private static final String RECEIVER_LINK_TAG_PREFIX = "cbs-receiver";
+    private static final String LINK_TYPE = "cbs";
```

Comment:
protonj seems to be a more native-ish library for amqp; the .NET amqp library gives you a cbs link directly as a predefined type!

Reply:
I do find it interesting that the .NET library does that for you. It probably speaks to how much of a Microsoft thing CBS auth is

Reply:
AMQP itself does not define CBS authentication so much as it allows for services to design systems like it, after all